python/packages/ai/teams/ai/ai.py

```diff
@@ -76,3 +100,221 @@ def __call__(func: Callable[[TurnContext, StateT, Any, str], Awaitable[bool]]):
+            print("hello world!")
+            return True
+
+        # Pass a function to this method
+        app.ai.function()(hello_world)
+        ```
+
+        #### Args:
+        - `name`: The name of the action `Default: Function Name`
+        - `allow_overrides`: If it should throw an error when duplicates
+        are found `Default: False`
+        """
+
+        def __call__(func: Callable[[TurnContext, StateT], Awaitable[str]]):
+            func_name = name
+
+            if not func_name:
+                func_name = func.__name__
+
+            self._options.prompt_manager.add_function(
+                name=func_name, handler=func, allow_overrides=allow_overrides
+            )
+            return func
+
+        return __call__
+
+    async def chain(
+        self, context: TurnContext, state: StateT, prompt: Union[str, PromptTemplate]
+    ) -> bool:
+        """
+        Chains into another prompt and executes the plan that is returned.
+
+        This method is used to chain into another prompt. It will call the prompt manager to
+        get the plan for the prompt and then execute the plan. The return value indicates whether
+        that plan was completely executed or not, and can be used to make decisions about whether
+        the outer plan should continue executing.
+        """
+
+        rendered = await self._options.prompt_manager.render_prompt(context, state, prompt)
+        plan = None
+
+        if self._review_prompt:
+            plan = await self._review_prompt(context, state, rendered)
+
+        if not plan:
+            plan = await self._options.planner.generate_plan(
+                context, state, rendered, history_options=self._options.history
+            )
+
+        if self._review_plan:
+            plan = await self._review_plan(context, state, plan)
+
+        plan_ready = self._actions.get(ActionTypes.PLAN_READY)
+
+        if not plan_ready:
+            return False
+
+        is_ok = await plan_ready.invoke(context, state, plan)
+
+        if not is_ok:
+            return False
+
+        if self._options.history.track_history:
+            state.conversation.value.history.add(
+                "user", state.temp.value.input.strip(), self._options.history.max_turns * 2
+            )
+
+            if self._options.history.assistant_history_type == "text":
+                text = "\n".join(
+                    map(
+                        lambda cmd: PredictedSayCommand.from_dict(cmd.__dict__).response,
+                        filter(lambda cmd: cmd.type == "SAY", plan.commands),
+                    )
+                )
+
+                state.conversation.value.history.add("assistant", text)
+            else:
+                state.conversation.value.history.add("assistant", plan.json())
+
+        for cmd in plan.commands:
+            if cmd.type == "DO":
+                do_cmd = PredictedDoCommand.from_dict(cmd.__dict__)
```

Comment:
Maybe we can use `isinstance` function to detect the command type and finish the type inferring? The type of `plan.commands` can be changed to `Union[PredictedDoCommand, PredictedSayCommand]`. I personally not a fan of current method.

Reply:
good idea I didn't think of that 👍